File: code/library/src/main/java/org/fbme/lib/iec61499/descriptors/FBTypeDescriptor.java
package org.fbme.lib.iec61499.descriptors;

import org.fbme.lib.common.Declaration;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public interface FBTypeDescriptor {

    @NotNull String getTypeName();

    @Nullable Declaration getDeclaration();

    @NotNull List<FBPortDescriptor> getEventInputPorts();

    @NotNull List<FBPortDescriptor> getEventOutputPorts();

    @NotNull List<FBPortDescriptor> getDataInputPorts();

    @NotNull List<FBPortDescriptor> getDataOutputPorts();

    @NotNull List<FBPortDescriptor> getSocketPorts();

    @NotNull List<FBPortDescriptor> getPlugPorts();

    @NotNull List<Integer> getAssociatedVariablesForInputEvent(int eventNumber);

    @NotNull List<Integer> getAssociatedVariablesForOutputEvent(int eventNumber);
}

File: code/library/src/main/kotlin/org/fbme/lib/iec61499/descriptors/FBTypeDescriptor.kt
package org.fbme.lib.iec61499.descriptors

import org.fbme.lib.common.Declaration

interface FBTypeDescriptor {
    val typeName: String
    val declaration: Declaration?
    val eventInputPorts: List<FBPortDescriptor>
    val eventOutputPorts: List<FBPortDescriptor>
    val dataInputPorts: List<FBPortDescriptor>
    val dataOutputPorts: List<FBPortDescriptor>
    val socketPorts: List<FBPortDescriptor>
    val plugPorts: List<FBPortDescriptor>
    fun getAssociatedVariablesForInputEvent(eventNumber: Int): List<Int>
    fun getAssociatedVariablesForOutputEvent(eventNumber: Int): List<Int>
}

data class FBTypeDescriptorImpl(
    override val typeName: String,
    override val declaration: Declaration?,
    override val eventInputPorts: List<FBPortDescriptor> = listOf(),
    override val eventOutputPorts: List<FBPortDescriptor> = listOf(),
    override val dataInputPorts: List<FBPortDescriptor> = listOf(),
    override val dataOutputPorts: List<FBPortDescriptor> = listOf(),
    override val socketPorts: List<FBPortDescriptor> = listOf(),
    override val plugPorts: List<FBPortDescriptor> = listOf(),
) : FBTypeDescriptor {

    override fun getAssociatedVariablesForInputEvent(eventNumber: Int): List<Int> = listOf()

    override fun getAssociatedVariablesForOutputEvent(eventNumber: Int): List<Int> = listOf()
}